controllers/networkfence_controller.go

```diff
@@ -113,37 +118,68 @@ func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
-	// check if the networkfence object is getting deleted, if yes, then send
-	// UnfenceCluster request to the driver.
-	// After the UnfenceCluster request is processed, the finalizer will be removed
-	// so that the networkfence object can be deleted gracefully.
-	if !nwFence.GetDeletionTimestamp().IsZero() {
+	nf := NetworkFenceInstance{
+		ctx:              ctx,
+		reconciler:       r,
+		logger:           logger,
+		instance:         nwFence,
+		controllerClient: client,
+	}
+
+	// check if the networkfence object is getting deleted and handle it.
+	if !nf.instance.GetDeletionTimestamp().IsZero() {
 		if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
 
-			err := r.unfenceClusterNetwork(ctx, nwFence, client, logger)
+			err := nf.removeFinalizerFromNetworkFence()
```

Comment:
(consider) Delete should ensure state is as desired before deletion. IOW, a fence resource request that did not succeed and is being deleted, may leave the system fenced or unfenced (and vice versa).

For example, if the resource status for the current generation being deleted reports desired state (due to a prior reconcile) then proceed with deletion, else retry the operation as we could have timed out in the past, and the system would be in an inconsistent state (some parts maybe fenced/unfenced as per the spec.FenceState).

This is not as critical as a CreateVolume or snapshot request that may leak resources if deleted prematurely (IOW, not ensuring existing system state), we may still get into corner cases where this maybe leave the system inconsistent.

Reply:
Yes. Perhaps it can be done in a separate PR? The current way of handling is similar to how it was being handled before before this PR. With current changes, it may not be exactly same as how it was before this PR as before for delete request, the reconciler would attempt unfencing. However, even with that, the issue you mentioned would exist as it did not check the current state of the world before attempting unefnce operation as part of deletion.